game/game/playercontrol.cpp

```diff
@@ -405,15 +415,21 @@ void PlayerControl::marvinK(uint64_t dt) {
   if (w == nullptr || w->player() == nullptr)
     return;
 
-  auto& pl = *w->player();
+  auto& pl  = *w->player();
   auto  pos = pl.position();
   float rot = pl.rotationRad();
   float s = std::sin(rot), c = std::cos(rot);
-
   Tempest::Vec3 dp(s, 0.0f, -c);
-  pos += dp * 6000 * float(dt) / 1000.f;
+  dp *= 6000 * float(dt) / 1000.f;
 
   pl.clearState(false);
+  if (pl.isSwim()) {
+    auto waterRay = pl.world().physic()->waterRay(pos);
+    auto res      = pl.world().physic()->ray(pos,{pos.x,20000,pos.z});
```

Comment:
does marvin+K behaves the same in original?

Reply:
Vanilla marvinK in diving state considers `rotationYRad` so it's like a generalized marvinF8. Whenever there's an obstacle in the path the game tries to port the character upwards. This behavior is not consistent, sometimes nothing happens. I replicated basically the same behavior in the new commit, because sometimes no collision is detected.